Several fixes to box inference

Linyxus opened the pull request:
Inspired by the formalism, this PR changes the box adaptation logic to fix a bunch of problems.

The main fixes:

- *Correctly charging the capture set of the type being adapted.* For example, the following code snippet should fail:
```scala
type Id[X] = [T] -> (op: X -> T) -> T
val x: Id[{io} Cap] = ???
x(cap => cap.use())  // error
```
Denote `cap => cap.use()` as `f`. The expected type is `(□ {io} Cap) -> Int`, while the actual type is `({io} Cap) -> Int`. The function should be adapted as `(x: □ {io} Cap) => f(unbox {io} x)`, whose capture set is `{io}` because of the unbox.

- *Stop the newly-captured variables from escaping a box*. Specifically, assuming that `T` is a function type, if the actual type is `T` while the expected type is `□ U`, the variables captured when adapting `T` to `U` will not get charged to the outer environment. For example, the following code should compile:
```scala
type Box[X] = X
type Id[X] = Box[X] -> Unit
type Op[X] = Unit -> Box[X]
val f: Unit -> ({io} Cap) -> Unit = ???
val g: {} Op[{io} Id[{io} Cap]] = f
```
The expected type of `f` in the last statement is `{} Unit -> □ {io} (□ {io} Cap) -> Unit`. The function would be adapted as `(x: Unit) => let g = (y: □ {io} Cap) => f(unbox {io} y) in box g`. Note that although an `unbox` is inferenced, outermost capture set is still empty due to the `box` we insert, which conforms to the expectation.

Besides, it also adds the support to adapting polymorphic function types.

odersky commented on compiler/src/dotty/tools/dotc/cc/CheckCaptures.scala:
```diff
@@ -593,25 +605,124 @@ class CheckCaptures extends Recheck, SymTransformer:
+
+          val (parent1, cs1) = parent match {
+            case actual @ AppliedType(tycon, args) if defn.isNonRefinedFunction(actual) =>
+              val (parent1, cs1) = adaptFun(parent, args.init, args.last, expected, covariant, insertBox,
```
Consider a different name for `cs1` since it's not the same value as the outer `cs1`

odersky replied:
(same in the instances below)

odersky reviewed (changes requested):
Overall it looks really good!

Linyxus commented:
Thanks for your comment! I have addressed the reviews and the main changes include:

- Named args and renaming to make the code easier to understand.
- Remove the incorrect usage of `markSolved` and fix a limitation in `TypeComparer` to make things work.
- Improve the way of destructing expected type in `adaptFun` and `adaptTypeFun` to make them match better with each other and have a clear rationale.

odersky reviewed (approved):
LGTM!
